fancybox/fancybox.php

```diff
@@ -28,27 +28,43 @@ function fancybox_footer(App $a, string &$str)
 	DI::page()->registerFooterScript(__DIR__ . '/asset/fancybox/fancybox.config.js');
 }
 
-function fancybox_render(App $a, array &$b)
-{
-	$matches = [];
-	$pattern = '#<div class="body-attach">.*?</div>#s';
-	$gallery = 'gallery';
-	if (array_key_exists('item', $b)) {
-		$item = $b['item'];
-		if (array_key_exists('uri-id', $item)) {
-			$gallery = $gallery . '-' . $item['uri-id'];
-		}
-	}
-	$html = $b['html'];
-	while (preg_match($pattern, $html, $matches, PREG_OFFSET_CAPTURE)) {
-		if (is_array($matches)) {
-			$matches = $matches[0];
-		}
-		$part     = $matches[0];
-		$replaced = str_replace('<a href', '<a data-fancybox="' . $gallery . '" href', $part);
-		$replaced = str_replace('<div class="body-attach"', '<div class="body-attach done"', $replaced);
-		$html     = str_replace($part, $replaced, $html);
+function fancybox_render($a, array &$b){
+	$gallery = 'gallery-' . $b['item']['uri-id'] ?? random_int(1000000, 10000000);
+
+	// prevent urls in <div class="type-link"> to be replaced
+	$b['html'] = preg_replace_callback(
+		'#<div class="type-link">.*?</div>#s',
+		function ($matches) use ($gallery) {
+			return str_replace('<a href', '<a data-nofancybox="" href', $matches[0]);
+		},
+		$b['html']
+	);
```

Comment:
This can be done with an existing Friendica method called `Util\Strings::performWithEscapedBlocks` with the same regular expression you came up with.

It would look like this (untested):
```php
	$b['html'] = \Friendica\Util\Strings::performWithEscapedBlocks(
		$b['html'],
		'#<div class="type-link">.*?</div>#s',
		function ($text) use ($gallery) {
			// This processes images inlined in posts
			// Frio / Vier hooks für lightbox are un-hooked in fancybox-config.js. So this works for them, too!
			//if (!in_array($a->getCurrentTheme(),['vier','frio']))
			// normal post inline linked images
			$text = preg_replace(
				'#<a[^>]*href="([^"]*)"[^>]*>(<img[^>]*src="[^"]*"[^>]*>)</a>#',
				'<a data-fancybox="' . $gallery . '" href="$1">$2</a>',
				$text
			);

			// Local content images attached:
			return preg_replace_callback(
				'#<div class="body-attach">.*?</div>#s',
				function ($matches) use ($gallery) {
					return str_replace('<a href', '<a data-fancybox="' . $gallery . '" href', $matches[0]);
				},
				$text
			);
		}
	);
```
Notice the change from `preg_replace_callback` to `preg_replace` since we don't need to care about `no-fancybox` elements.

Reply:
Ah.. I see in the comment description of that function in code.. Hmm.. Have to check that..

Reply:
Were you able to test?

Reply:
Yes I tested this. It worked like intended. So I think the (not yet known to me) method does what it is supposed to do.
Img URLs in "type-link" divs are not touched. :-)

Reply:
I ran into the same exact use case you did and decided to make my future life and the future life of everyone who contributes to the project easier!